better lxd support

mkg20001 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Make nixOS container run better in lxd

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mkg20001 committed:
systemd: disable systemd-udev-trigger.service in containers

mkg20001 committed:
virtualisation.lxc-container: overwrite tarball for lxd

mkg20001 committed:
virtualisation.lxc: indent config

mkg20001 committed:
virtualisation.lxc: add support for lxd templates

mkg20001 committed:
virtualisation.lxc: use openFirewall instead of allowedTCPPorts

mkg20001 committed:
virtualisation.lxc: split meta and root

mkg20001 commented:
What is the equivalent to adding this in nixOS?

systemd.extraConfig doesn't seem to be right (still getting errors in unprivileged). I can't use environment.etc because systemd/service is managed by the generateUnits function (and adding it there results in a build error as it's .source to a read-only path)

```
/etc/systemd/system/service.d/lxc.conf

[Service]
ProtectProc=default
ProtectControlGroups=no
ProtectKernelTunables=no
```

mkg20001 committed:
virtualisation.lxc: add the overrides from distrobuilder

mkg20001 committed:
release.lxdImage: add lxdImage to hydra

andir commented on nixos/modules/virtualisation/lxc-container.nix:
```diff
@@ -1,26 +1,137 @@
+    };
+  };
+
+  toYAML = name: attrs: pkgs.runCommandNoCC name {
```
Why does it have to be YAML? if the tool is fine with JSON we can skip the local build and the requirement of having remarshal locally. I find it annoying very frequently with other modules.

andir replied:
Also if you want to do it this way remove the indirection through `writeFile` and use `passAsFile` instead. https://nixos.org/manual/nix/stable/#sec-advanced-attributes

mkg20001 replied:
I'd have to see if they support json, so far I've only seen yaml for metadata used

flokli replied:
YAML is a superset of JSON, so if they support parsing yaml, giving json should be fine.

mkg20001 replied:
This is based on the assumption the same code can parse both JSON and YAML, which is not the case.

andir commented on nixos/modules/virtualisation/lxc-container.nix:
```diff
@@ -1,26 +1,137 @@
-  # Allow ssh connections
-  networking.firewall.allowedTCPPorts = [ 22 ];
+    # Allow ssh connections
+    services.openssh.openFirewall = mkDefault true;
```
Isn't this already the default anyway?

mkg20001 replied:
Not sure. previously it was having allowedTCPPorts here, so I switched that over to openFirewall.

mkg20001 replied:
Seems enabled by default, removed

andir commented on nixos/modules/virtualisation/lxc-container.nix:
```diff
@@ -1,26 +1,137 @@
+          symlink = "none";
+        }
+        {
+          object = pkgs.stdenv;
```
Really the whole stdenv? This includes a compiler etc. Did you mean to populate the system with coreutils etc. instead?

mkg20001 replied:
Docker profile also included it in their tarball, but I can remove it if that's better

andir replied:
Interesting. Not sure why we would prefer a full environment with compiler instead of a minimal "more pure" system. I would say it was probably not discussed back then?

mkg20001 replied:
The idea was being able to rebuild offline, as rebuild did download some stdenv things when running nixos-rebuild switch. Do you think this is a feature that's worth having?

andir commented on nixos/modules/virtualisation/lxc-container.nix:
```diff
@@ -1,26 +1,137 @@
+  options = {
+    virtualisation.lxc = {
+      templates = mkOption {
+        description = "Templates for LXD";
```
Could you provide an example value? And a link to relevant LXD docs if those exist?

mkg20001 replied:
One template is in nixos/maintainers/script/lxd/lxd-image.nix. I'll add docs.

mkg20001 replied:
I couldn't find any docs on the format, but I added some example templates.

andir reviewed (changes requested):
This doesn't look too bad.

From what I can say this is lacking:

* A NixOS VM test
* Release notes
* Documentation on what is expected to be in a template or how custom templates look like.

flokli commented on nixos/maintainers/scripts/lxd/lxd-image.nix:
```diff
@@ -0,0 +1,34 @@
+    fi
+  '';
+
+  # Make lxc exec work properly
```
can lxc fall back to a `/bin/sh` for `lxc exec` if it doesn't find `/bin/bash`? That'd probably even be applicable upstream.

mkg20001 replied:
The problem is with the sh not having the path set properly actually. Plus a reduced developer experience because of that and sh's rather minimal featureset.

flokli replied:
Well, `/bin/sh` IS a symlink to `bash-interactive` on NixOS, so there's no worse developer experience. And even for non-NixOS, giving a stripped down `/bin/sh` instead of failing entirely might be preferrable.

mkg20001 replied:
Please note that while /bin/sh is indeed a symlink to bash, bash was programmed to behave like sh when argv[0] is sh (when the link is called "sh"). Meaning it will only ever have the features posix sh has enabled, unless it gets called as /bin/bash (or literally anything that is not sh). If it doesn't load the profile then that is a problem.

Ofc if you know of a way to get sh to load the profile in this particular circumstance I'm all ears. Just without $PATH set correctly the shell isn't exactly useful

mkg20001 committed:
virtualisation.lxc: apply changes from review

mkg20001 committed:
virtualisation.lxc: add default for .templates

mkg20001 committed:
virtualisation.lxc: replace /sbin/init on switch

in the future maybe a proper shim-bootloader could be used that supports 
switching out /sbin/init (so nixos-rebuild boot would just switch the 
link)

mkg20001 commented:
When trying to add packages to the container, I see most of them being rebuilt and some fail with odd errors

Does anyone know if this is in any way related to the container profile?

```
error: builder for '/nix/store/qyfvscign772q8xjk5vmrcviz1mbfcjc-gtk+3-3.24.27.drv' failed with exit code 1;
       last 10 log lines:
       > Run-time dependency xkbcommon found: YES 1.2.1
       > Run-time dependency iso-codes found: YES 4.6.0
       > Run-time dependency tracker-sparql-3.0 found: YES 3.0.3
       > Did not find CMake 'cmake'
       > Found CMake: NO
       > Run-time dependency cairo-xlib found: NO (tried pkgconfig)
       >
       > meson.build:521:97: ERROR: Dependency "cairo-xlib" not found, tried pkgconfig
       >
       > A full log can be found at /build/gtk+-3.24.27/build/meson-logs/meson-log.txt
       For full logs, run 'nix log /nix/store/qyfvscign772q8xjk5vmrcviz1mbfcjc-gtk+3-3.24.27.drv'.
error: 1 dependencies of derivation '/nix/store/3pc1q2h65j126nm6mllkn0l133nyngmh-stoken-0.92.drv' failed to build
error: 1 dependencies of derivation '/nix/store/qqzgvqx4wdfqjc2n73902q3k6dc21arv-openconnect-8.10.drv' failed to build
error: 1 dependencies of derivation '/nix/store/63faz5p4nh8qdiw2fxglc5zq6pp89bg9-NetworkManager-openconnect-1.2.6.drv' failed to build
error: 1 dependencies of derivation '/nix/store/j2c4qzmiqyysl3j5gxzs2jrajjsxjbq0-fix-paths.patch.drv' failed to build
error: 1 dependencies of derivation '/nix/store/gw9p3zk32bflkqvbnmvhj7issa22cwvp-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/9nicjjjyn2yz24dwpi9bd2q4yik5yvfr-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/j7d6pdl625zjrqb7fx32mayyirkpbq45-nixos-system-helium-21.05.20210430.ff9e292.drv' failed to build
```

mkg20001 committed:
lxdImage: enable docs & xlibs in standalone image

mkg20001 commented:
Ah I solved it, environment.noXlibs. (profiles/docker-profile imports profiles/minimal)

Seems it causes a full rebuild of most things for very little benefit. Anything against disabling that option in lxd?

mweinelt commented:
@GrahamcOfBorg eval

mkg20001 commented:
Does it make sense to split this into smaller PRs so it's easier to merge?

lblasc commented:
Looking forward seeing this merged, @mkg20001 thanks! From the CI perspective only release note formatting needs to be adjusted, is there anything else left to fix ?

mkg20001 committed:
release-notes: add LXD support

mkg20001 committed:
lxdImage: split from docker profile, use generators.toYAML

mkg20001 committed:
nixosTest.lxdImage: add lxdImage test

mkg20001 commented:
@GrahamcOfBorg test lxd lxd-image lxd-nftables

mkg20001 commented:
@GrahamcOfBorg test lxd-image

mkg20001 commented:
Seems like lxd-image is passing, just the other test is broken

Merging it now
